Add file/event command

4e6 opened the pull request:
### Pull Request Description

closes #541 

Changelog:
- add `WatcherAdapter` watches directory and executes callback on file events
- add `PathWatcher` actor manages `WatcherAdapter`, handles errors and processes file system events from the watcher
- add `ReceivesTreeUpdatesHandler` actor handles capability requests
- add `file/event` notification message
- add `receivesTreeUpdates` capability

### Important Notes

`file/event` notification message returns `Path` instead of `FileSystemObject`.

``` typescript
{
  path: Path;
  kind: FileEventKind;
}
```

To construct the `FileSystemObject` we need to know it's type, i.e. is it a file, or directory, etc.
When we receive `Deleted` event, we can't resolve the path because at that moment it does not exist.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventWatcherApi.scala:
```diff
@@ -0,0 +1,68 @@
+        case DirectoryChangeEvent.EventType.OVERFLOW => None
+      }
+  }
+
```
This class defines API is can't be responsible for conversion between different APIs. Please move it to the FileEventWatcher.

lolczak replied:
It introduces coupling between low-level lib. and API. The usage of low-level lib. should be encapsulated by the FileEventWatcher. The FileEventWatcher leaks details that it is supposed to abstract away.

4e6 replied:
It indeed is an abstraction over the watcher library. It hides the implementation details. You are probably confused by the name, I'll remove the Api suffix.

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventWatcherApi.scala:
```diff
@@ -0,0 +1,68 @@
+    def from(event: DirectoryChangeEvent): Option[WatcherEvent] =
+      EventType
+        .from(event.eventType())
+        .map(WatcherEvent(event.path(), _))
```
It's again coupling between low-level lib. and our API.

4e6 replied:
It's not our public API. I will rename the class to avoid confusion.

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventRegistry.scala:
```diff
@@ -0,0 +1,295 @@
+      } else {
+        log.error(s"Unable to find a client for $msg")
+        manager ! FileEventManagerProtocol.UnwatchPath
+      }
```
The FileEventManager should be responsible for delivering events to a subscriber. Please pass to the FileEventManager the client handle it works for.

4e6 replied:
Currently (see interaction scheme in the description), FileEventManager doesn't know anything about the client and the registry and only replies to the sender. There are three isolated layers right now, and I don't see benefits in changing that.

lolczak replied:
The simplicity and lower latency is a benefit.

4e6 replied:
Updated. Now manager sends events directly to the client

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventManager.scala:
```diff
@@ -0,0 +1,87 @@
+  * @param config configuration
+  * @param exec executor of file system effects
+  */
+final class FileEventManager(config: Config, exec: Exec[BlockingIO])
```
It should subscribe to `ClientDiconnected` events and stop the manager when the subscriber dies.

4e6 replied:
Handling of `ClientDisconnected` messages is implemented.

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventRegistry.scala:
```diff
@@ -0,0 +1,295 @@
+  * @param config configuration
+  * @param exec executor of file system events
+  */
+final class FileEventRegistry(config: Config, exec: Exec[BlockingIO])
```
This actor introduces a lot of accidental complexity due to the chosen design. It should only forward requests to managers and create them if not exists. It shouldn't be responsible for complex routing responses and building a few maps to achieve that. It can be simply done by forwarding requests to managers and the manager shall reply to the right sender. Please see how BufferRegistry is implemented.

4e6 replied:
First, by adding arbitrary connections between actors you also introduce complexity. Imagine the graph (the actor system), where all nodes (actors) are interconnected. It's harder to reason about, compared to the layered system with fewer connections. What I'm trying to say, is that it's a tradeoff. By connecting the manager and the client, you will decrease the complexity of the registry, but increase the overall complexity of the system. 

Second, even if we connect the manager and the client, we can't completely remove routing from the registry. Think about the capability/release request. We still need to store the info about the clients and managers to forward the message to the right manager.

lolczak replied:
You don't add a connection to the actor but an observer as part of its state. You have poor naming for the FileEventManger. It is not a manger. This actor has a Subject role from the [Observer](https://en.wikipedia.org/wiki/Observer_pattern) design pattern (GOF). It's a design of the observer pattern that the subject holds a reference to observers that listen to the subject's updates.  Your layered design didn't remove the connection between participants (subject-observer). You moved this natural connection to a different layer adding accidental complexity that didn't come from the problem but come from the solution you chose. Moving this interdependency to the subject removes the whole layer of routing of both events and replies. Your design also decreases cohesion because the elements that belong together, because one concept is spread to many classes. When you move this natural connection as described in the design pattern, you will simplify the registry, and the subject will have the same complexity. So, it will be easier to reason about both the registry and the subject. 

Another issue is that the routing in the registry is implemented in the wrong way. The routing should be done by the path, not the client id. Multiple clients can observe one path. There should be one subject per path containing any number of observer objects.

lolczak replied:
Moreover, there is an intrinsic coupling between a subject and an observer. You introduced the registry, which acts as a mediator between those participants. This way, you moved this inherent coupling from the subject to the mediator, additionally introducing another artificial coupling between the subject and the mediator. In effect, we have more coupling (observer-mediator, mediator-subject) than it comes from the problem. I suggest using Occam razor to get rid of unnecessary coupling and concepts and keep it simple.

lolczak commented on engine/language-server/src/test/scala/org/enso/languageserver/websocket/FileEventRegistryTest.scala:
```diff
@@ -0,0 +1,232 @@
+              "message": "Service error"
+            }
+          }
+          """)
```
File not found should be returned.

4e6 replied:
Now we return `FileNotFound` error here

lolczak commented on engine/language-server/src/test/scala/org/enso/languageserver/websocket/FileEventRegistryTest.scala:
```diff
@@ -0,0 +1,232 @@
+              "message": "Service error"
+            }
+          }
+          """)
```
The more meaningful error should be returned.

4e6 replied:
Now we return `CapabilityNotAcquired` error here

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventManager.scala:
```diff
@@ -0,0 +1,87 @@
+  private def initializedStage(
+    root: File,
+    base: Path,
+    replyTo: ActorRef
```
ok

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventManager.scala:
```diff
@@ -0,0 +1,87 @@
+
+  import FileEventManagerProtocol._
+
+  private var fileWatcher: FileEventWatcher = _
```
done

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventManager.scala:
```diff
@@ -0,0 +1,87 @@
+          val watcherResult =
+            Try(FileEventWatcher.build(pathToWatch, self ! _))
+              .fold(resultFailure, resultSuccess)
+              .map { watcher =>
```
End up using ZIO

lolczak replied:
You have to include import cats.implicits._ and code above should work.

lolczak replied:
IO is not needed here

lolczak replied:
If FileEventWatcher is effectful it should signal it through API.

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventWatcher.scala:
```diff
@@ -0,0 +1,52 @@
+  * @param root directory to watch
+  * @param callback that fires on the file system events
+  */
+final class FileEventWatcher(
```
I added exception handling

lolczak replied:
It should restart a programmed number of times when error occurs.

4e6 replied:
Now, when a maximum amount of restarts occurs, we force-release the capability and notify the client.

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEvent.scala:
```diff
@@ -0,0 +1,106 @@
+        case CodecType.Removed  => Right(Removed)
+        case CodecType.Modified => Right(Modified)
+      }
+    }
```
I didn't find this feature in Circe, but it seems the `enumeratum` library can do that. Do we really need to bring a new library just for this one case?

lolczak replied:
Yes, it the enumeratum contains this behavior.  You can include this lib. It's very easy, and it's not time-consuming.

4e6 replied:
I found an issue with `enumeratum`. It encodes ADT as strings, 

``` json
"kind" : "Added"
```

which is inconsistent with our encoding of `FileSystemObject`. It should be.

``` json
"kind" : {
  "type" : "Added"
}
```

For users, different encoding may be confusing because in `engine-services.md` we define both types the same way.

```typescript
type FileEventKind = Added | Removed | Modified;
```

```typescript
type FileSystemObject
  = Directory
  | DirectoryTruncated
  | SymlinkLoop
  | File
  | Other;
```

lolczak replied:
FileEventKind is defined as `type FileEventKind = Added | Removed | Modified;` it should be encoded as a String. Why do you need a wrapper around it?

lolczak replied:
If FileSystemObject is wrapped in this way. It also should be fixed.

4e6 replied:
`FileSystemObject` is encoded as a JSON object, and needs field `type` for decoding, i.e. to distinguish between different constructors.

``` scala
  case class Directory(name: String, path: Path) extends FileSystemObject
  case class DirectoryTruncated(name: String, path: Path) extends FileSystemObject
```

My logic is if `FileSystemObject` is encoded as a JSON object, then `FileEventKind` should be encoded the same way

lolczak replied:
So where do you see a link between FileSystemObject and FileEventKind, if it is a completely different class? The FileEventKind is a typical enum that should be encoded as constant. If CirceEnum encodes it this way we can assume that it is an industry-standard. If Occam razor and simplicity can't convince you, so maybe the [Twitter API](https://developer.twitter.com/en/docs/ads/general/overview/enums) cause that you weaken your resistance. This a json snippet from Twitter API, look how approval_status and industry_type are encoded :
```
   "data": {
       "name": "API McTestface",
       "business_name": null,
       "timezone": "America/Los_Angeles",
       "timezone_switch_at": "2016-07-21T07:00:00Z",
       "id": "18ce54d4x5t",
       "created_at": "2016-07-21T22:42:09Z",
       "salt": "54cb7b5a34183f77d82dd6d0f4329691",
       "updated_at": "2017-07-06T16:51:04Z",
       "industry_type": "TRAVEL",
       "business_id": null,
       "approval_status": "ACCEPTED",
       "deleted": false
     }
   }
```

4e6 replied:
I added `enumeratum` library.

4e6 commented:
@lolczak I addressed all comments except the one about bringing `enumeratum` library that didn't work for us, and the design thingy, on which we disagree.

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventManager.scala:
```diff
@@ -0,0 +1,180 @@
+        )
+      } else {
+        log.error("Hit maximum number of restarts", e)
+        subscriber ! FileEventError(e)
```
It should stop when can't recover from failure.

4e6 replied:
Updated. Not now the manager stops itself, and not by the registry.

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventRegistry.scala:
```diff
@@ -0,0 +1,333 @@
+  * @param config configuration
+  * @param exec executor of file system events
+  */
+final class FileEventRegistry(
```
It should monitor managers and remove them from the store when target stops. Otherwise, it can cause memory leak.

4e6 replied:
Updated. Registry monitors managers

4e6 commented:
@lolczak Please review again. I added `enumeratum` library, and now the manager sends events directly to the client.

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventManager.scala:
```diff
@@ -0,0 +1,184 @@
+    root: File,
+    base: Path,
+    subscriber: ActorRef,
+    client: ActorRef
```
The client is a subscriber (observer) of updates. You can send both replies and events and updates to subscribers directly. You don't need another layer of [adapters](https://en.wikipedia.org/wiki/Adapter_pattern). The FileEventWatcher is an adapter that hides low-level API of file watching capability. The one layer of adapters is enough. The FileEventManager can convert file watching messages directly to CapabilityProtocol and FileManagerProtocol.
The FileEventManager should be renamed to PathWatcher, and it should notify all clients that watch the path it was created for.

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/FileEventManager.scala:
```diff
@@ -0,0 +1,184 @@
+  private val errorHandler: Throwable => FileSystemFailure = {
+    case ex => GenericFileSystemFailure(ex.getMessage)
+  }
+}
```
Since the FileEventWatcher (I suggest renaming it to WatcherAdapter) performs effectful computations, the usage of the IO monad here is improper. The FileEventWatcher should have an IO monad in its signature. Thus, its methods should return IO[IOException, Unit], or you can map failures and return IO[FileSystemFailure, Unit].

4e6 commented:
ChangeLog:
- rename `FileEventWatcher` to `WatcherAdapter` and add `IO` to type signatures.
- rename `FileEventManager` to `PathWatcher` which now talks to the client
- rename `FileEventRegistry` to  `ReceivesTreeUpdatesHandler` that handles capability requests and forwards them to corresponding `PathWatcher`

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/PathWatcher.scala:
```diff
@@ -0,0 +1,225 @@
+    stopWatcher(): Unit
+  }
+
+  override def receive: Receive = uninitializedStage
```
Please implement the `unhandled` method.

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/filemanager/ReceivesTreeUpdatesHandler.scala:
```diff
@@ -0,0 +1,126 @@
+  * @param fs file system
+  * @param exec executor of file system events
+  */
+final class ReceivesTreeUpdatesHandler(
```
Implement the `unhandled` method.

lolczak commented on engine/language-server/src/main/scala/org/enso/languageserver/requesthandler/ReleaseCapabilityHandler.scala:
```diff
@@ -32,7 +36,7 @@ class ReleaseCapabilityHandler(
     case Request(ReleaseCapability, id, params: CapabilityRegistration) =>
-      capabilityRouter ! CapabilityProtocol.ReleaseCapability(client.id, params)
+      capabilityRouter ! CapabilityProtocol.ReleaseCapability(client, params)
       context.system.scheduler.scheduleOnce(timeout, self, RequestTimeout)
```
We should cancel this event when the reply comes on time. If you can add this feature to the rest of the handlers.
